gemma-rest/src/main/java/ubic/gemma/rest/DatasetsWebService.java

```diff
@@ -1211,24 +1215,45 @@ public Object getDatasetCellTypeAssignment(
             } else {
                 cta = found.iterator().next();
             }
-        } else if ( ctaName != null ) {
-            cta = singleCellExpressionExperimentService.getCellTypeAssignment( ee, qt, ctaName );
-            if ( cta == null ) {
-                throw new NotFoundException( "No cell type assignment with name " + ctaName + " found for " + ee.getShortName() + " and " + qt.getName() + "." );
+        } else if ( ctaIdentifier != null ) {
+            CellTypeAssignment cta2;
+            try {
+                Long ctaId = Long.parseLong( ctaIdentifier );
+                cta2 = singleCellExpressionExperimentService.getCellTypeAssignment( ee, qt, ctaId );
+                if ( cta2 == null ) {
+                    throw new NotFoundException( "No cell type assignment with ID " + ctaIdentifier + " found for " + ee.getShortName() + " and " + qt.getName() + "." );
+                }
+            } catch ( NumberFormatException e ) {
+                cta2 = singleCellExpressionExperimentService.getCellTypeAssignment( ee, qt, ctaIdentifier );
+                if ( cta2 == null ) {
+                    throw new NotFoundException( "No cell type assignment with name " + ctaIdentifier + " found for " + ee.getShortName() + " and " + qt.getName() + "." );
+                }
             }
+            cta = cta2;
         } else {
             cta = singleCellExpressionExperimentService.getPreferredCellTypeAssignment( ee, qt )
                     .orElseThrow( () -> new NotFoundException( "No preferred cell type assignment found for " + ee.getShortName() + " and " + qt.getName() + "." ) );
         }
-        MediaType negotiate = negotiate( headers, MediaType.APPLICATION_JSON_TYPE, TEXT_TAB_SEPARATED_VALUES_UTF8_TYPE );
-        if ( negotiate.equals( TEXT_TAB_SEPARATED_VALUES_UTF8_TYPE ) ) {
-            return ( StreamingOutput ) output -> {
-                CellLevelCharacteristicsWriter writer = new CellLevelCharacteristicsWriter();
-                writer.setUseBioAssayIds( useBioAssayId );
-                writer.write( cta, dimension, new OutputStreamWriter( output, StandardCharsets.UTF_8 ) );
-            };
+        if ( download ) {
+            return Response.ok( ( StreamingOutput ) output -> {
+                        CellLevelCharacteristicsWriter writer = new CellLevelCharacteristicsWriter();
+                        writer.setUseBioAssayIds( useBioAssayId );
+                        writer.write( cta, dimension, new OutputStreamWriter( new GZIPOutputStream( output ), StandardCharsets.UTF_8 ) );
+                    } )
+                    .header( "Content-Disposition", "attachment; filename=\"" + cta.getId() + "_" + ( cta.getName() != null ? FileTools.cleanForFileName( cta.getName() ) : "cell_type" ) + ".tsv.gz\"" )
```

Comment:
This logic could be moved in `ExpressionDataFileUtils` alongside other helper for generating data file names.